[Demo] Perturbative Gadgets for Variational Quantum Algorithms

SimonCichy opened the pull request:
**Title:** Perturbative Gadgets for Variational Quantum Algorithms

**Summary:**
A tutorial to explore the application of perturbative gadgets in variational quantum algorithms to delay the onset of cost-function-dependent barren plateaus. Substituting a global Hamiltonian with its gadgetized equivalent allows to train the algorithm on a local cost function and still finding the target minimum, thus avoiding the issue of cost-function-dependent barren plateaus.

**Relevant references:**
https://github.com/SimonCichy/barren-gadgets/tree/main/Pennylane-tutorial
https://arxiv.org/pdf/2210.03099.pdf

**Possible Drawbacks:**

**Related GitHub Issues:** #593

SimonCichy commented:
Of course, I did not pay attention when translating the changes from the clone to the fork. Hopefully the checks are successful now

josh146 commented:
Looking good so far! Thanks @SimonCichy. Note that the first builds of a PR here may take awhile, but once successful, subsequent builds are much faster :)

josh146 commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,332 @@
+
+In this tutorial, we will explore the application of perturbative gadgets in 
+variational quantum algorithms to outgo the issue of cost function dependent
+barren plateaus, as proposed in Ref.[#cichy2022]
```
```suggestion
barren plateaus, as proposed in Ref.[#cichy2022]_
```
This will ensure it automatically links to the Bibliography at the bottom

josh146 commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,332 @@
+Some context
+------------
+
+If you are not familiar yet with the concept of barren plateaus, I recomend you
+first check out these other tutorials: 
+- https://pennylane.ai/qml/demos/tutorial_barren_plateaus.html
+- https://pennylane.ai/qml/demos/tutorial_local_cost_functions.html  
```
```suggestion
If you are not familiar yet with the concept of barren plateaus, I recommend you
first check out the demonstrations on :doc:`barren plateaus </demos/tutorial_barren_plateaus>`
and :doc:`avoiding barren plateaus with local cost functions </demos/tutorial_local_cost_functions>`.
```

Using this style of local link ensure that the tutorial is not affected if we ever change the URL of the website :)

josh146 commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,332 @@
+acting on a total of :math:`8` qubits, that has :math:`3` terms, each of them being a 
+:math:`4`-body interaction. 
+
+#.. figure:: ../demonstrations/barren_gadgets/gadget-terms-tutorial.png
+#    :align: center
+#    :width: 90%
```
```suggestion
.. figure:: ../demonstrations/barren_gadgets/gadget-terms-tutorial.png
    :align: center
    :width: 90%
```

josh146 commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,332 @@
+# As ansatz, we will use a variation of the
+# `qml.SimplifiedTwoDesign <https://pennylane.readthedocs.io/en/latest/code/api/pennylane.SimplifiedTwoDesign.html#pennylane.SimplifiedTwoDesign>`_,
+# which was proposed in previous
+# works on cost function dependent barren plateaus [#cerezo2021].
```
```suggestion
# works on cost function dependent barren plateaus [#cerezo2021]_.
```

josh146 commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,332 @@
+The penalization (red) acts only on the auxiliary registers, penalizing each 
+qubit individually, while the perturbations couple target with auxiliary qubits.
+
+As shown in Ref. [#cichy2022], this construction results in a spectrum that is similar
```
```suggestion
As shown in Ref. [#cichy2022]_, this construction results in a spectrum that is similar
```

KetpuntoG commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,332 @@
+plt.plot(costs_lists["monitoring"])
+plt.legend(["training", "monitoring"])
+plt.show()
+
```
<img width="715" alt="Captura de Pantalla 2022-10-19 a las 13 59 22" src="https://user-images.githubusercontent.com/65235481/196769014-6ce3fdac-bf1d-4aff-959f-0b8be8b3dbbe.png">

In this case I think it is not necessary to show the prints, the graph gives all the necessary information of the final result :)

SimonCichy replied:
Sure! I'll remove it at the end. It's just helpful for me when checking that it runs as expected, to have a quick reaction and not waiting for the full run to know it works

KetpuntoG commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,332 @@
+
+import pennylane as qml
+from pennylane import numpy as np
+import matplotlib.pyplot as plt
```
It is more didactic to put the imports when you need them. I would put `matplotlib` library in the last cell and add the "seaborn" style to make it more elegant.

```python
import matplotlib.pyplot as pet
plt.style.use("seaborn")
```

KetpuntoG reviewed (commented):
Again, great job! 😊 Here are some first comments and suggestions about the demo to work on them

josh146 commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,332 @@
+#
+#    Cerezo, M., Sone, A., Volkoff, T. et al.
+#    "Cost function dependent barren plateaus in shallow parametrized quantum circuits." `Nat Commun 12, 1791
+#    <https://doi.org/10.1038/s41467-021-21728-w>`__, 2021.
```
@SimonCichy you can also add an author bio here :) It looks like this:

```suggestion
#    <https://doi.org/10.1038/s41467-021-21728-w>`__, 2021.
#
# About the author
# ----------------
#
# .. bio:: Simon Cichy
#    :photo: ../_static/authors/simon_cichy.jpg
# 
#    Simon is...
```

You can also feel free to add any links (e.g., to personal websites) in your bio if you would like to!

SimonCichy commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,332 @@
+# `qml.Hamiltonian <https://pennylane.readthedocs.io/en/stable/code/api/pennylane.Hamiltonian.html>`_
+# class
+
+term1 = qml.operation.Tensor(qml.PauliX(0), qml.PauliX(1), qml.PauliY(2), qml.PauliZ(3))
```
Does it generate the same kind of object? I am aware one can write it in that way, but I have the feeling that at some point when working on my code it made a difference whether I used one or the other. It has been a few months, so I'm probably wrong about that, but was it always the case, that generating the Hamiltonian as you propose you could still obtain the Hcomp.ops and Hcomp.coeffs?

SimonCichy replied:
Found the issue: for this example, where the Hamiltonian has several terms, it's fine and everything works out. However, in the next example where the Hamiltonian has only one term the object is a Tensor, and does not have the attribute ops (which I use in the gadgetization)

SimonCichy replied:
Solved (kind of): I had missed the factor 1 you added in the second example. With the factor it indeed becomes a Hamiltonian object. I had never realized that, that's probably why I used the Hamiltonian class throughout my code.

KetpuntoG replied:
That's right, it may seem counterintuitive but a PauliZ gate by itself we do not consider a Hamiltonian in Pennylane. When we multiply it by 1 we interpret it as a Hamiltonian

SimonCichy commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,332 @@
+acting on a total of :math:`8` qubits, that has :math:`3` terms, each of them being a 
+:math:`4`-body interaction. 
+
+.. figure:: ../demonstrations/barren_gadgets/gadget-terms-tutorial.png
```
Thanks!

KetpuntoG replied:
I don't know if it will be very difficult but in the image you would have to change "s" for "i" if I'm not mistaken.

SimonCichy commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,332 @@
+# So, let us construct the two Hamiltonians of interest
+
+term1 = qml.operation.Tensor(qml.PauliX(0), qml.PauliY(1), qml.PauliZ(2), qml.PauliZ(3))
+Hcomp = qml.Hamiltonian([1], [term1])
```
Oh, I see, the '1 *' was not decorative. With this added Hcomp is indeed a Hamiltonian object

josh146 replied:
Note that we are in the process of updating/improving our operators in PennyLane. this is a long term project, but it will start to get better/more consistent with each release 😆 

Soon the 1 will no longer be needed!

KetpuntoG commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,378 @@
+
+.. math:: H^\text{aux}_i = \sum_{j=1}^k |1\rangle \langle 1|_{i,j} = \sum_{j=1}^k \frac{1}{2}(\mathbb{I} - Z_{i,j})
+
+while the perturbation part implements one of the operators in the Pauli word
```
```suggestion
On the other hand, the perturbation part implements one of the operators in the Pauli word

```

KetpuntoG commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,378 @@
+# from a user-given target Hamiltonian in an automated way. 
+# For those who want to check the inner workings,
+# you can find the code here (
+# :file:`../demonstrations/barren_gadgets/barren_gadgets.py` 
```
```suggestion
```

KetpuntoG commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,378 @@
+# The ``PerturbativeGadget`` class allows to generate the gadget Hamiltonian
+# from a user-given target Hamiltonian in an automated way. 
+# For those who want to check the inner workings,
+# you can find the code here (
```
```suggestion
# you can find the code `here <https://pennylane.ai/qml/demos/barren_gadgets/barren_gadgets.py>`__

```

KetpuntoG replied:
I think this solves the problem

josh146 replied:
Lets see if the `:file:` -> `:download:` fix solves it 🤞

KetpuntoG commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,378 @@
+# For those who want to check the inner workings,
+# you can find the code here (
+# :file:`../demonstrations/barren_gadgets/barren_gadgets.py` 
+# ).
```
```suggestion
# .

```

KetpuntoG commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,378 @@
+# :file:`../demonstrations/barren_gadgets/barren_gadgets.py` 
+# ).
+# The functions ``get_parameter_shape``, ``generate_random_gate_sequence`` and
+# ``build_ansatz`` (for the details:
```
```suggestion
# ``build_ansatz`` (more details `here <https://pennylane.ai/qml/demos/barren_gadgets/layered_ansatz.py>`__

```

josh146 replied:
Lets see if the :file: -> :download: fix solves it 🤞

KetpuntoG replied:
Hey @josh146 , If we click in the link we get this xml file

```xml
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AccessDenied</Code><Message>Access Denied</Message><RequestId>GK9B1D0DDGCGDV5Y</RequestId><HostId>eMLLUZDSVw1cXaAUfnwGAwwFMsNNre/X1yHKPZDeK3z2RIgTLCpxnQdh/8/ivsLDp2uOzxQMTXg=</HostId></Error>

```

Maybe it won't work in the current deployment but it will once merged?

josh146 replied:
I just tried to fix this!

KetpuntoG commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,378 @@
+# ).
+# The functions ``get_parameter_shape``, ``generate_random_gate_sequence`` and
+# ``build_ansatz`` (for the details:
+# :file:`../demonstrations/barren_gadgets/layered ansatz.py` 
```
```suggestion
```

josh146 replied:
```suggestion
# :download:`../demonstrations/barren_gadgets/layered ansatz.py` 
```

KetpuntoG commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,378 @@
+#
+# First we construct our target Hamiltonian in Pennylane.
+# For this, we use the
+# `qml.Hamiltonian <https://pennylane.readthedocs.io/en/stable/code/api/pennylane.Hamiltonian.html>`_
```
This comment is a reminder to me to update this link. We use to reference our classes in other ways

KetpuntoG commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,378 @@
+# We indeed have a Hamiltonian composed of two terms, with the expected Pauli
+# words.
+# Next, we can construct the corresponding gadget Hamiltonian.
+# Using the class PerturbativeGadgets, we can automatedly
```
```suggestion
# Using the class `PerturbativeGadgets`, we can automatedly

```

KetpuntoG commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,378 @@
+# the training cost function, the optimizer with its step size, and the number of
+# optimization steps and the device to run the circuit on.
+# As ansatz, we will use a variation of the
+# `qml.SimplifiedTwoDesign <https://pennylane.readthedocs.io/en/latest/code/api/pennylane.SimplifiedTwoDesign.html#pennylane.SimplifiedTwoDesign>`_,
```
same here, I will update this link

KetpuntoG commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,378 @@
+##############################################################################
+# Since our example of target Hamiltonian is a single Pauli string, we know
+# without needing any training that it has only :math:`\pm 1` eigenvalues.
+# It is a very simple example. but we see that the training of our circuit using
```
```suggestion
# It is a very simple example but we see that the training of our circuit using

```

KetpuntoG reviewed (commented):
Here are a few other comments :)
It's usually a good practice when working on the review, to start accepting the suggestions when there is the option (if you agree 😄) and then work on the comments. 
On the other hand, If you have already worked on a particular comment, it would be good for me if you put 👍 to be sure that you have worked on it
And of course, if you have any questions, feel free to ask

josh146 commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,378 @@
+# from a user-given target Hamiltonian in an automated way. 
+# For those who want to check the inner workings,
+# you can find the code here (
+# :file:`../demonstrations/barren_gadgets/barren_gadgets.py` 
```
```suggestion
# :download:`../demonstrations/barren_gadgets/barren_gadgets.py` 
```

josh146 replied:
@KetpuntoG sorry I think this was my fault - I think this will fix it!

SimonCichy commented:
Thanks @KetpuntoG for your comment on good practices, that helps a lot. Should I also add the changes I make as suggestions? I'll have to check how to do that

josh146 commented:
> Should I also add the changes I make as suggestions?

@SimonCichy completely up to you, whichever you prefer :) Feel free to commit/push locally using git. For small/one liner changes, you could also use the Suggestions functionality on GitHub

josh146 commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,379 @@
+# from a user-given target Hamiltonian in an automated way. 
+# For those who want to check the inner workings,
+# you can find the code here (
+# :download:`../demonstrations/barren_gadgets/barren_gadgets.py` 
```
```suggestion
# :download:`barren_gadgets.py <../demonstrations/barren_gadgets/barren_gadgets.py>` 
```

josh146 commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,379 @@
+# ).
+# The functions ``get_parameter_shape``, ``generate_random_gate_sequence`` and
+# ``build_ansatz`` (for the details:
+# :download:`../demonstrations/barren_gadgets/layered_ansatz.py` 
```
```suggestion
# :download:`layered_ansatz.py <../demonstrations/barren_gadgets/layered_ansatz.py>` 
```

KetpuntoG commented:
Hey @SimonCichy , it would help me for follow the review if you mark comments with a "👍" if you have already worked with them. 

<img width="840" alt="Captura de Pantalla 2022-10-26 a las 9 16 30" src="https://user-images.githubusercontent.com/65235481/198035758-ef9c5630-ed5c-4733-8eed-c193ec1a3669.png">

On the other hand, if you have already worked with all of them, you can let me know by requesting a new review here:

<img width="384" alt="Captura de Pantalla 2022-10-26 a las 9 18 40" src="https://user-images.githubusercontent.com/65235481/198036309-1842a1e7-4852-4ad5-8444-c35a5ace0c5d.png">

Simply click on the symbol 🔄

SimonCichy commented:
Hi @KetpuntoG, I will try to remember reacting to your comments when working on them. There are a few on which I had worked but not notified you yet (mainly on adding explanations), because I thought I would re-read later to see if my explanations where clear. Then I could not get back for some days due to other things. I will get back to it soon, and mark your comments once I am happy with it

SimonCichy commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,332 @@
+import pennylane as qml
+from pennylane import numpy as np
+import matplotlib.pyplot as plt
+from barren_gadgets.barren_gadgets import PerturbativeGadgets
```
If you consider the explanations sufficient, and now that we have the download links, this could be resolved, right?

KetpuntoG commented:
Great, I look forward to the re-review notification! 😄

SimonCichy commented:
I seem to have a bit of an issue: I have changed quite a few things according to the comments, but the changes are not reflected in the latest deployments. Not sure if there is something wrong with what I added, but I don't know how to debug it either.

KetpuntoG commented:
> @KetpuntoG, @josh146, I seem to have a bit of an issue: I have changed quite a few things according to the comments, but the changes are not reflected in the latest deployments. Not sure if there is something wrong with what I added, but I don't know how to debug it either.

Hey @SimonCichy !This may be due to the fact that you are loading the demo from cache. To fix it, copy the link and open it in an incognito window :)

SimonCichy commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,379 @@
+##############################################################################
+# Finally, we will use two cost functions.
+# For each we create a
+# `QNode <https://docs.pennylane.ai/en/stable/code/api/pennylane.QNode.html>`_.
```
And here I added a reference to the QNode, maybe it is not necessary, but since I was adding references to the other classes I did this one too

SimonCichy commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,379 @@
+# .. bio:: Simon Cichy
+#    :photo: ../_static/authors/simon_cichy.jpg
+# 
+#    `Simon <https://simoncichy.github.io/about.html/>`__ is a graduate in quantum engineering from ETH Zürich. He is interested in quantum machine learning and near-term quantum algorithms in general. 
```
And that link does not seem to work, I am not sure why yet

KetpuntoG commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,379 @@
+
+
+qml.draw_mpl(display_circuit)(weights)
+
```
to prevent the output from appearing, add `plt.show()` to the end of the cell.
(import matplotlib before if you haven't done so)

KetpuntoG reviewed (commented):
Awesome job! For my part, with this last comments, the demo would be ready 😊 we will give a second review before publication. The next reviewer will comment if there are new things to modify

SimonCichy commented:
Great! Happy to see it slowly taking shape :) I believe I have implemented all the comments on my side, the only things I see left are the comments from you @KetpuntoG about updating the links to reference pennylane classes (Hamiltonian x2, QNode and SimplifiedTwoDesign). Also, for some reason the link in the author bio does not seem to work when writing it like the other links.

SimonCichy commented:
> Also, for some reason the link in the author bio does not seem to work when writing it like the other links.
All the other author bios including links to personal websites have them in plain text, so maybe it can't be done as is now?

josh146 commented:
> All the other author bios including links to personal websites have them in plain text, so maybe it can't be done as is now?

Ah, this might be a bug in our bio setup. Feel free to just use plain text for now, I will investigate and attempt to fix this. But this shouldn't block this demo :)

CatalinaAlbornoz commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,380 @@
+For each of the terms :math:`h_1`, :math:`h_2` and :math:`h_3` we add :math:`4` auxiliary qubits.
+In the end, our gadget Hamiltonian thus acts on :math:`8+3\cdot 4 = 20` qubits.
+The penalization (red) acts only on the auxiliary registers, penalizing each 
+qubit individually, while the perturbations couple target with auxiliary qubits.
```
```suggestion
qubit individually, while the perturbations couple the target with the auxiliary qubits.
```

CatalinaAlbornoz commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,380 @@
+barren plateaus than the global cost, so it is more trainable.
+As a result, one can mitigate the onset of cost function dependent barren 
+plateaus by substituting the global cost function with the resulting gadget
+and using that for training instead. That is what we will do in the following
```
```suggestion
and using that for training instead. That is what we will do in the rest
```

CatalinaAlbornoz commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,380 @@
+"""
+
+##############################################################################
+# First, a few imports. Pennylane and numpy of course, and additionally a few
```
```suggestion
# First, a few imports. PennyLane and NumPy of course, and additionally a few
```

CatalinaAlbornoz commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,380 @@
+# functions specific to our tutorial. 
+# The ``PerturbativeGadget`` class allows to generate the gadget Hamiltonian
+# from a user-given target Hamiltonian in an automated way. 
+# For those who want to check the inner workings,
```
```suggestion
# For those who want to check its inner workings,
```

CatalinaAlbornoz commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,380 @@
+# The ``PerturbativeGadget`` class allows to generate the gadget Hamiltonian
+# from a user-given target Hamiltonian in an automated way. 
+# For those who want to check the inner workings,
+# you can find the code here (
```
```suggestion
# you can find the code here (
```

CatalinaAlbornoz commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,380 @@
+# from a user-given target Hamiltonian in an automated way. 
+# For those who want to check the inner workings,
+# you can find the code here (
+# :download:`barren_gadgets.py <../demonstrations/barren_gadgets/barren_gadgets.py>` 
+# ).
```
```suggestion
# :download:`barren_gadgets.py <../demonstrations/barren_gadgets/barren_gadgets.py>`
# ).
```

CatalinaAlbornoz commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,380 @@
+# The last 8 are the perturbation. They are a little scambled, but one can
+# recognize the 8 Paulis from the target Hamiltonian on the qubits ``0`` to 
+# ``3`` and the cyclic pairwise :math:`X` structure on the auxiliaries.
+# Indeed, there are :math:`(X_4X_5, X_5X_6, X_6X_7, X_7X_4)` and
```
```suggestion
# Indeed, they are :math:`(X_4X_5, X_5X_6, X_6X_7, X_7X_4)` and
```

CatalinaAlbornoz commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,380 @@
+
+##############################################################################
+# Then we need to set up our variational quantum algorithm.
+# That is we choose a circuit ansatz with randomly initialized weights,
```
```suggestion
# That is, we choose a circuit ansatz with randomly initialized weights,
```

CatalinaAlbornoz reviewed (dismissed):
Very interesting demo Simon! 

I suggested a few changes but they should be easy to commit.

SimonCichy commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,380 @@
+Barren plateaus refer to the phenomenon where the gradients of the cost function
+decay exponentially with the size of the problem. Essentially, the cost 
+landscape becomes flat, with exception of some small regions, e.g., around
+the minimum. That is a problem because increasing the resolution on a quantum 
```
Sorry for this abuse of language, I would have to check the proper definitions of resolution, accuracy and precision. I mean that to obtain a good expectation value of a quantity one has to perform many measurements to overcome the randomness of the shot noise

SimonCichy replied:
Ok, the use of the word resolution was clearly wrong (that is fixed by the hardware, so does not enter in the consideration here). According to this visualization, I mean increase the precision (in the assumption that it is already accurate)
![image](https://user-images.githubusercontent.com/70891818/201912717-cb6afc3c-a48b-404c-8320-ca0c2fec7eb2.png)

CatalinaAlbornoz replied:
Awesome! Yes, precision seems like the right word to use!

SimonCichy commented:
> Very interesting demo Simon!
> 
> I suggested a few changes but they should be easy to commit.

Thank you a lot! Thanks for the detailed review, and sorry for all the typos that slipped through my fingers. About your comments, I had accepted the easy suggested changes, and will go in the following days over those that need a bit of thinking ;)

SimonCichy commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,380 @@
+# .. bio:: Simon Cichy
+#    :photo: ../_static/authors/simon_cichy.jpg
+# 
+#    Simon is a graduate in quantum engineering from ETH Zürich. He is interested in quantum machine learning and near-term quantum algorithms in general. More on his personal website https://simoncichy.github.io/about
```
Well, that depends a bit on how long this this process still takes. I have finished all the requirements for the master, but still waiting for the paperwork to be done. And I will start a PhD, but same there, it is not yet formally the case. I will change it to graduate student and update it later if the PR is still open by then

CatalinaAlbornoz replied:
Sounds good!

SimonCichy commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,375 @@
+landscape becomes flat, with exception of some small regions, e.g., around
+the minimum. That is a problem because increasing the resolution on a quantum 
+computer comes at the cost of more measurements, and an exponential number of 
+measurements would render the algorithm impractical.
```
```suggestion
the minimum. That is a problem because increasing the precision on a quantum 
computer comes at the cost of more measurements.
Shot noise adds some randomness in the computed expectation value
and the smaller the difference to observe, the more measurements
are needed. An exponential number of 
measurements would render the algorithm impractical.

```

SimonCichy commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,375 @@
+decay exponentially with the size of the problem. Essentially, the cost 
+landscape becomes flat, with exception of some small regions, e.g., around
+the minimum. That is a problem because increasing the resolution on a quantum 
+computer comes at the cost of more measurements, and an exponential number of 
```
```suggestion
computer comes at the cost of more measurements.
Shot noise adds some randomness in the computed expectation value
and the smaller the difference to observe, the more measurements
are needed. An exponential number of 
measurements would render the algorithm impractical.
```

SimonCichy commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,375 @@
+Barren plateaus refer to the phenomenon where the gradients of the cost function
+decay exponentially with the size of the problem. Essentially, the cost 
+landscape becomes flat, with exception of some small regions, e.g., around
+the minimum. That is a problem because increasing the resolution on a quantum 
+computer comes at the cost of more measurements, and an exponential number of 
+measurements would render the algorithm impractical.
```
```suggestion
the minimum. That is a problem because increasing the precision on a quantum 
computer comes at the cost of more measurements.
Shot noise adds some randomness in the computed expectation value
and the smaller the difference to observe, the more measurements
are needed. An exponential number of 
measurements would render the algorithm impractical.

```

isaacdevlugt replied:
```suggestion
the minimum. That is a problem because increasing the resolution of the cost function requires more measurements from the quantum device, and an exponential number of 
measurements would render the algorithm impractical.

```

SimonCichy replied:
@CatalinaAlbornoz what do you think? Since these suggestions also relate to your doubt about the use of the word "resolution"
> What do you mean by resolution here? Do you mean accuracy? Or precision? Or something else? It may be worth describing what this means in this context.

CatalinaAlbornoz replied:
Hi @SimonCichy, I believe that Isaac didn't see your suggestion because you hadn't committed it. 

Maybe a combination of the 2 is better. Something like:

That is a problem because increasing the precision while using a quantum 
computer requires more measurements from the quantum device.
Shot noise adds some randomness in the computed expectation value
and the smaller the difference to observe, the more measurements
are needed. An exponential number of 
measurements would render the algorithm impractical.

josh146 commented:
@KetpuntoG: it turns out the `:download:` issue is a bug, currently waiting here on a fix :)

rashidnhm commented:
Hi all, the download issue should now be resolved, if you are able to pull in the latest master to your branch, the download links to begin working.

KetpuntoG commented:
> Hi all, the download issue should now be resolved, if you are able to pull in the latest master to your branch, the download links to begin working.

Awesome! Thanks @rashidnhm @josh146 , now it is working for me 💪

KetpuntoG commented:
Hey @SimonCichy , did you make all the requested changes already? 
I just want to make sure before I give it another look :)

SimonCichy commented:
> Hey @SimonCichy , did you make all the requested changes already? I just want to make sure before I give it another look :)

@KetpuntoG I believe I did! I have accepted most of the proposed changes, suggested improvements in some places and there is one where I discuss my choice of formulation. I am not aware of any that is pending of me right now

isaacdevlugt commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,374 @@
+act on a small subset of the qubits in the register.
+We want to explore this topic further and learn about one possible mitigation
+strategy.  
+Thinking about VQE applications, let us consider cost functions that are 
```
```suggestion
Thinking about Variational Quantum Eigensolver (VQE) applications, let us consider cost functions that are 

```

isaacdevlugt commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,374 @@
+
+Here :math:`|00\ldots 0\rangle` is our initial state, 
+:math:`V(\theta)` is the circuit ansatz and :math:`H` the Hamiltonian
+to minimize.  
```
```suggestion
whose expectation value we need to minimize.  

```

isaacdevlugt commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,374 @@
+barren plateaus than the global cost, so it is more trainable.
+As a result, one can mitigate the onset of cost-function-dependent barren 
+plateaus by substituting the global cost function with the resulting gadget
+and using that for training instead. That is what we will do in the rest
```
```suggestion
and use that for training instead. That is what we will do in the rest

```

isaacdevlugt commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,374 @@
+"""
+
+##############################################################################
+# First, a few imports. PennyLane and NumPy of course, and additionally a few
```
```suggestion
# First, a few imports. PennyLane and NumPy of course, and a few

```

isaacdevlugt commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,374 @@
+# The ``PerturbativeGadget`` class allows the user to generate the gadget Hamiltonian
+# from a user-given target Hamiltonian in an automated way. 
+# For those who want to check its inner workings,
+# you can find the code here (
+# :download:`barren_gadgets.py </demonstrations/barren_gadgets/barren_gadgets.py>`
+# ).
```
```suggestion
# you can find the code here:
# :download:`barren_gadgets.py </demonstrations/barren_gadgets/barren_gadgets.py>`.

```

isaacdevlugt commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,374 @@
+# you can find the code here (
+# :download:`barren_gadgets.py </demonstrations/barren_gadgets/barren_gadgets.py>`
+# ).
+# The functions ``get_parameter_shape``, ``generate_random_gate_sequence`` and
```
```suggestion
# The functions ``get_parameter_shape``, ``generate_random_gate_sequence``, and

```

isaacdevlugt commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,374 @@
+np.random.seed(3)
+
+##############################################################################
+# Now, let's take the example given above
```
```suggestion
# Now, let's take the example given above:

```

isaacdevlugt commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,374 @@
+# First, we construct our target Hamiltonian in PennyLane.
+# For this, we use the
+# `qml.Hamiltonian <https://pennylane.readthedocs.io/en/stable/code/api/pennylane.Hamiltonian.html>`_
+# class
```
```suggestion
# class.

```

isaacdevlugt commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,374 @@
+print(H_target)
+
+##############################################################################
+# We indeed have a Hamiltonian composed of two terms, with the expected Pauli
```
```suggestion
# We indeed have a Hamiltonian composed of two terms with the expected Pauli

```

isaacdevlugt commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,374 @@
+# -----------------------------------
+# Now that we have a little intuition on how the gadget Hamiltonian construction
+# works, we will use it to train.
+# Classical simulation of qubit systems is expensive, so we will simplify further
```
```suggestion
# Classical simulations of qubit systems are expensive, so we will simplify further

```

isaacdevlugt commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,374 @@
+# Classical simulation of qubit systems is expensive, so we will simplify further
+# to a target Hamiltonian with a single term, and show that using the
+# gadget Hamiltonian for training allows us to minimize the target Hamiltonian.
+# So, let us construct the two Hamiltonians of interest
```
```suggestion
# So, let us construct the two Hamiltonians of interest.

```

isaacdevlugt commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,374 @@
+##############################################################################
+# Then we need to set up our variational quantum algorithm.
+# That is, we choose a circuit ansatz with randomly initialized weights,
+# the training cost function, the optimizer with its step size, the number of
```
```suggestion
# the cost function, the optimizer with its step size, the number of

```

isaacdevlugt commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,374 @@
+# Then we need to set up our variational quantum algorithm.
+# That is, we choose a circuit ansatz with randomly initialized weights,
+# the training cost function, the optimizer with its step size, the number of
+# optimization steps and the device to run the circuit on.
```
```suggestion
# optimization steps, and the device to run the circuit on.

```

isaacdevlugt commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,374 @@
+# That is, we choose a circuit ansatz with randomly initialized weights,
+# the training cost function, the optimizer with its step size, the number of
+# optimization steps and the device to run the circuit on.
+# As ansatz, we will use a variation of the
```
```suggestion
# For an ansatz, we will use a variation of the

```

isaacdevlugt commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,374 @@
+plt.show()
+
+##############################################################################
+# Now we build the circuit for our actual experiment
```
```suggestion
# Now we build the circuit for our actual experiment.

```

isaacdevlugt commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,374 @@
+# Finally, we will use two cost functions and create a
+# `QNode <https://docs.pennylane.ai/en/stable/code/api/pennylane.QNode.html>`_ for each.
+# The first cost function, the training cost, is the loss function of the optimization.
+# That's the one the gradient descent will try to minimize.
```
Not needed IMO 

```suggestion
```

isaacdevlugt commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,374 @@
+# `QNode <https://docs.pennylane.ai/en/stable/code/api/pennylane.QNode.html>`_ for each.
+# The first cost function, the training cost, is the loss function of the optimization.
+# That's the one the gradient descent will try to minimize.
+# For the training, we use the gadget Hamiltonian and
+# then we also define a monitoring cost, based on the target Hamiltonian.
+# We will evaluate its value at each iteration for monitoring purposes, but it
+# will not be used in the optimization.
```
```suggestion
# For the training, we use the gadget Hamiltonian. To ensure
# that our gadget optimization is proceeding as intended, 
# we also define another cost function based on the target Hamiltonian.
# We will evaluate its value at each iteration for monitoring purposes, but it
# will not be used in the optimization.

```

isaacdevlugt commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,374 @@
+plt.show()
+
+##############################################################################
+# Since our example of target Hamiltonian is a single Pauli string, we know
```
```suggestion
# Since our example target Hamiltonian is a single Pauli string, we know

```

isaacdevlugt commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,374 @@
+##############################################################################
+# Since our example of target Hamiltonian is a single Pauli string, we know
+# without needing any training that it has only :math:`\pm 1` eigenvalues.
+# It is a very simple example but we see that the training of our circuit using
```
```suggestion
# It is a very simple example, but we see that the training of our circuit using

```

isaacdevlugt commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,374 @@
+# 
+# Now that you have an idea of how you can use perturbative gadgets in 
+# variational quantum algorithms, you can try applying them to more complex
+# problems! However, be reminded of the exponential scaling of classical 
+# simulations of quantum systems, adding linearly many auxiliary qubits
+# quickly becomes a lot.
```
```suggestion
# problems! However, be aware of the exponential scaling of classical 
# simulations of quantum systems; adding linearly many auxiliary qubits
# quickly becomes hard to simulate.

```

isaacdevlugt reviewed (changes requested):
Some grammar comments, but will approve once addressed!

KetpuntoG commented:
Hey @SimonCichy , we plan to publish the demo next Tuesday 
Once you correct the last changes, it will be ready 😄
You could also share me your Twitter username for the publication

KetpuntoG commented:
> Hey @SimonCichy , we plan to publish the demo next Tuesday
> Once you correct the last changes, it will be ready 😄
> You could also share me your Twitter username for the publication

Also, if you could add a hook to your first paragraph (the two lines you have) that would be amazing. Something that would make the user want to read the whole demo 💪

SimonCichy commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,374 @@
+:math:`V(\theta)` is the circuit ansatz and :math:`H` the Hamiltonian
+to minimize.  
+
+In some cases, it is easy to find a local cost function which can substitute a global one with the same ground state. 
```
@isaacdevlugt sorry, but this rephrasing sounds weird to me. If you don't like the "which",  I understand using "that" instead but the "for" does not feel right to me there. Or alternatively using "that can act as a substitute for"

SimonCichy commented:
Hey @KetpuntoG! Sure, I'll get to it, and it seems that the changes requested by @isaacdevlugt are quick to go through. There are a few from @CatalinaAlbornoz however about which I was not sure, and left some comments. Those are not settled yet.

SimonCichy commented:
> You could also share me your Twitter username for the publication
@sdcichy

SimonCichy commented:
> Don't forget to add a hook in the first two lines you've written. Apart from summarizing what they are going to see, we are trying to make the user curious about the demo

I added a few sentences, but I am happy to hear suggestions if you have better ones! It is oversimplifying things and I don't know how much hype it builds, but I don't want to sell what we did for more than it is

CatalinaAlbornoz commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,380 @@
+This means that minimizing the gadget Hamiltonian, if reaching its global
+minimum, the resulting state will be close to the global minimum of 
+:math:`H^\text{target}` too.
+Since it is a local cost function, it is better behaved with respect to 
```
My comment was more about the "it is better behaved" statement. It is unclear what it is. Maybe something like the following is best. What do you think?

Since it is a local cost function, it will not exhibit barren plateaus, making it more trainable than the global cost function.

CatalinaAlbornoz commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,412 @@
+first check out the demonstrations on :doc:`barren plateaus </demos/tutorial_barren_plateaus>`
+and :doc:`avoiding barren plateaus with local cost functions </demos/tutorial_local_cost_functions>`.
+
+As presented in the second aforementioned demo, barren plateaus are more severe when using global
+
+cost functions compared to local ones. 
```
There are some weird spaces like this one appearing in several places. This requires a good review to make sure we catch them all.

CatalinaAlbornoz commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,412 @@
+
+As a result, one can mitigate the onset of cost-function-dependent barren 
+plateaus by substituting the global cost function with the resulting gadget
+and use that for training instead. That is what we will do in the rest
+
+of this tutorial.
```
```suggestion
and using that for training instead. That is what we will do in the rest of this tutorial.
```

josh146 commented on demonstrations/tutorial_barren_gadgets.py:
```diff
@@ -0,0 +1,400 @@
+# constructions with similar properties.  
+# Also, the complete code with explanations on how to reproduce the 
+# figures from the paper can be found in 
+# `this repository <https://github.com/SimonCichy/barren-gadgets>`_.
+##############################################################################
+# References
```
```suggestion
# `this repository <https://github.com/SimonCichy/barren-gadgets>`_.
#
# References
```

KetpuntoG reviewed (approved):
Awesome demo! Thank you very much @SimonCichy 🚀

SimonCichy commented:
Thank you guys a lot!! And sorry I could not be more responsive on the last part... But really happy this is out now 😃
